src/paths/voronoi.c

```diff
@@ -0,0 +1,423 @@
+    igraph_vector_fill(mindist, IGRAPH_INFINITY);
+
+    /* Loop through all generator points and compute shortest paths to all other vertices.
+     * As we go, we keep track of the shortest distance from any generator to each vertex
+     * in 'mindist'. If we find that the distance from the current generator to a vertex
+     * is shorter than what was recorded so far in 'mindist', we update 'mindist' and
+     * assign that vertex to the current generator.
+     */
+    for (igraph_integer_t i=0; i < no_of_generators; i++) {
+        igraph_integer_t g = VECTOR(*generators)[i];
+
+        IGRAPH_ALLOW_INTERRUPTION();
+
+        /* BFS-based unweighted shortest path implementation */
+
+        igraph_dqueue_int_clear(&q);
+
+        VECTOR(already_counted)[g] = i+1;
+        IGRAPH_CHECK(igraph_dqueue_int_push(&q, g));
+        IGRAPH_CHECK(igraph_dqueue_int_push(&q, 0));
+
+        while (!igraph_dqueue_int_empty(&q)) {
+            igraph_integer_t vid  = igraph_dqueue_int_pop(&q);
+            igraph_integer_t dist = igraph_dqueue_int_pop(&q);
+
+            /* Attention! This must be igraph_real_t, not igraph_integer_t
+             * because later it will be compared with another igraph_real_t
+             * whose value may be infinite. */
+            igraph_real_t md = VECTOR(*mindist)[vid];
+
+            if (dist > md) {
+                /* This vertex is reachable at a shorter distance from
+                 * another generator. Thus all its descendants in the shortest
+                 * path tree are also reachable at a shorter distance from the
+                 * other generator than from the current one. Therefore
+                 * we do not need to search further from here. */
+                continue;
+            } else if (dist < md) {
+                /* This vertex is closest to the current generator so far.
+                 * Assign it to the current partition. */
+                VECTOR(*mindist)[vid] = dist;
+                VECTOR(*membership)[vid] = i;
+                if (tiebreaker == IGRAPH_VORONOI_RANDOM) {
+                    VECTOR(tie_count)[vid] = 1;
+                }
+            } else { /* md == dist, we have a tie */
+                switch (tiebreaker) {
+                case IGRAPH_VORONOI_FIRST:
+                    /* Never replace existing generator assignment. */
+                    break;
+                case IGRAPH_VORONOI_LAST:
+                    /* Always replace existing generator assignment. */
+                    VECTOR(*membership)[vid] = i;
+                    break;
+                case IGRAPH_VORONOI_RANDOM:
+                    /* We replace the membership assignment with probability 1/k upon
+                     * encountering the kth same-distance generator. This ensures
+                     * that one of these generators is selected uniformly at random. */
+                    VECTOR(tie_count)[vid]++;
+                    RNG_BEGIN();
```

Comment:
Let's move these out of the loop; in R, this operation involves copying the entire state vector of the RNG from R-space to C-space (or the other way round).

Reply:
I'm assuming we don't care if `RNG_END()` is not called upon error, right?

Reply:
RIght, that's a mess, ideally it should be called, but the only use-case for `RNG_BEGIN()` and `RNG_END()` is R's peculiarity and the worst thing that can happen there is that the R-side RNG does not get updated with the state changes on the C side.

In the long term, `RNG_BEGIN()` and `RNG_END()` should be eliminated from the C core and moved entirely to the R side, after annotating our functions in `functions.yaml` so we know which ones use the RNG and which ones do not.